tokenizers/src/models/unigram/trainer.rs

```diff
@@ -425,22 +454,22 @@ impl UnigramTrainer {
                 }
                 (objs, ntokens, expected)
             })
-            .collect();
-
-        let mut expected: Vec<f64> = vec![0.0; model.len()];
-        let mut objs: f64 = 0.0;
-        let mut ntokens: u32 = 0;
-
-        for (lobjs, lntokens, lexpected) in collected {
-            objs += lobjs;
-            ntokens += lntokens;
-            expected
-                .iter_mut()
-                .zip(lexpected)
-                .for_each(|(global_el, local_el)| *global_el += local_el)
-        }
+            .reduce(
+                || (0.0, 0, vec![0.0; model.len()]),
+                |(objs, ntokens, expected), (lobjs, lntokens, lexpected)| {
+                    (
+                        objs + lobjs,
+                        ntokens + lntokens,
+                        expected
+                            .iter()
+                            .zip(lexpected)
+                            .map(|(global_el, local_el)| global_el + local_el)
+                            .collect(),
```

Comment:
`iter_mut()` + `for_each` makes sure there are not reallocations.

`.collect()` is like `.clone()`, avoid if possible.

It might be optimized away but I would rather not count on it.

Reply:
Hum but you need to set that variable to `mut` no?

Reply:
Probably, but if the reduce is not itself parallel, then it's ok to have `global` be mutable.